File: schedmgr/descriptor.go
// SPDX-License-Identifier: Apache-2.0
package main

import "strings"

const pgKeychainPrefix = "pg:keychain:"

// parseDBDescriptor classifies the --db value. The canonical Postgres descriptor is
// "pg:keychain:<service>" — a NON-secret reference resolved against the OS keychain at runtime
// (spec §3.3; the crontab line must never carry the DSN). Anything else is a SQLite file path.
func parseDBDescriptor(s string) (isPg bool, pgService string) {
	if strings.HasPrefix(s, pgKeychainPrefix) {
		return true, strings.TrimPrefix(s, pgKeychainPrefix)
	}
	return false, ""
}

File: src/main/scheduler/descriptor.ts
// SPDX-License-Identifier: Apache-2.0
import type { BackendConfigFile } from '../db/backendConfig'

/** The non-secret schedmgr DB descriptor baked into cron lines / Task actions. Postgres uses a
 *  `pg:keychain:<service>` reference — schedmgr resolves the DSN from the OS keychain at runtime, so
 *  the secret never touches the crontab (spec §3.3). SQLite uses the plain db file path. */
export function schedmgrDbDescriptor(cfg: BackendConfigFile, sqlitePath: string): string {
  return cfg.backend === 'postgres' && cfg.pgService ? `pg:keychain:${cfg.pgService}` : sqlitePath
}
